### PR TITLE
Add negative_prompt_field support to SAM2 models

### PR DESCRIPTION
**Pull request by CharlesCNorton**
What changes are proposed in this pull request?

This PR adds support for negative prompts to SAM2 image and video models, enabling fine-grained segmentation control by excluding unwanted regions from segmentation masks.

Key changes:
  - Add negative_prompt_field parameter to both SegmentAnything2ImageModel and SegmentAnything2VideoModel
  - Convert negative bounding boxes to 4 corner points with label=0 (SAM2's negative prompt convention)
  - Support negative keypoints with label=0 for video model
  - Handle tensor shape mismatches when batching multiple boxes with negative prompts
  - Fully backward compatible - negative_prompt_field is optional

Implementation details:
  - Image model: Override predict_all() and modify _forward_pass_boxes() to extract and process negative prompts
  - Video model: Extend _get_field(), _forward_pass_boxes(), and _forward_pass_points() to handle negative prompts per frame
  - Process boxes individually when both multiple positive and negative prompts exist to avoid tensor shape issues
  - Fall back to batch processing when possible for optimal performance

How is this patch tested? If it is not, please explain why.

This patch includes comprehensive testing:

Unit tests (8 tests total):
  - Field extraction with and without frames. prefix
  - Negative box to corner point conversion logic
  - Negative keypoint label assignment
  - Backward compatibility without negative prompts
  - Support for both Detections and Keypoints as negative prompt types

Manual verification:
  - Tested on misc segmentation images in CL and FO app, for ex. dog/cat segmentation showing 6,414 pixels (2.7%) successfully excluded when using negative prompts
  - Verified docstring examples work correctly for both image and video models
  - Confirmed backward compatibility with existing code

Release Notes

SAM2 models now support negative prompts for refined segmentation control. Users can specify a negative_prompt_field parameter when applying SAM2 models to exclude unwanted regions from segmentation masks. This enables more precise segmentation by providing both positive prompts (regions to segment) and negative prompts (regions to exclude). Works with both image and video SAM2 models using Detections or Keypoints.

What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [X] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

* **New Features**
  * Negative-prompt support for image and video segmentation: per-image or per-frame negative examples (boxes, points, keypoints) can be supplied to subtract regions from masks and influence final segmentations.

* **Tests**
  * Added comprehensive unit tests for negative-prompt parsing, coordinate/label conversion, per-frame behavior, backward compatibility, and detection/keypoint format support.

* **Documentation**
  * Added usage examples demonstrating negative-prompt workflows for image and video.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: fiftyone/utils/sam2.py [hunk 1]
```diff
@@ -165,13 +186,46 @@ def _load_model(self, config):
...
+                negative_field = negative_field[len("frames."):]
+
+        if negative_field and samples is not None:
+            negative_prompts = []
+            for sample in samples:
+                try:
+                    value = sample.get_field(negative_field)
+                except AttributeError:
+                    value = None
+                negative_prompts.append(value)
```
Comment (lorenzoriano) [inline fiftyone/utils/sam2.py:262]: Under which condition would you see None here? And is it ok to append None instead of empty string ? Would a warning be ok here?

File: fiftyone/utils/sam2.py [hunk 2]
```diff
@@ -423,6 +527,19 @@ def _forward_pass_boxes(self, video_reader, sample):
...
+                if self._curr_negative_prompts and out_frame_idx < len(self._curr_negative_prompts):
+                    neg_frame_detections = self._curr_negative_prompts[out_frame_idx]
+                    if neg_frame_detections and len(neg_frame_detections.detections) > 0:
+                        for neg_det in neg_frame_detections.detections:
+                            neg_box = neg_det.bounding_box
+                            nx1 = int(neg_box[0] * self._curr_frame_width)
+                            ny1 = int(neg_box[1] * self._curr_frame_height)
+                            nx2 = int((neg_box[0] + neg_box[2]) * self._curr_frame_width)
+                            ny2 = int((neg_box[1] + neg_box[3]) * self._curr_frame_height)
+                            mask[ny1:ny2, nx1:nx2] = 0
```
[the remark this answers is not in the document]
Comment (lorenzoriano) [reply fiftyone/utils/sam2.py]: agree it'll be good to have those

File: fiftyone/utils/sam2.py [hunk 3]
```diff
@@ -205,8 +259,24 @@ def _forward_pass_boxes(self, imgs):
...
                 multimask_output=False,
             )
+
+            # Subtract negative prompt regions from masks
+            if self._curr_negative_prompts and idx < len(self._curr_negative_prompts):
+                neg_detections = self._curr_negative_prompts[idx]
+                if neg_detections and len(neg_detections.detections) > 0:
+                    for neg_det in neg_detections.detections:
+                        neg_box = neg_det.bounding_box
+                        nx1 = max(0, int(neg_box[0] * w))
```
Comment (lorenzoriano) [inline fiftyone/utils/sam2.py]: can you use fosam._to_abs_boxes() ?

File: fiftyone/utils/sam2.py [hunk 4]
```diff
@@ -423,6 +527,19 @@ def _forward_pass_boxes(self, video_reader, sample):
...
+
+                # Subtract negative prompt regions from mask
+                if self._curr_negative_prompts and out_frame_idx < len(self._curr_negative_prompts):
+                    neg_frame_detections = self._curr_negative_prompts[out_frame_idx]
+                    if neg_frame_detections and len(neg_frame_detections.detections) > 0:
+                        for neg_det in neg_frame_detections.detections:
+                            neg_box = neg_det.bounding_box
+                            nx1 = int(neg_box[0] * self._curr_frame_width)
+                            ny1 = int(neg_box[1] * self._curr_frame_height)
+                            nx2 = int((neg_box[0] + neg_box[2]) * self._curr_frame_width)
```
Comment (lorenzoriano) [inline fiftyone/utils/sam2.py]: can you use fosam._to_abs_boxes() ?

File: fiftyone/utils/sam2.py (hunk 3 above)
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py]: Done.

File: fiftyone/utils/sam2.py (hunk 4 above)
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py]: Also done!

**Comment by CharlesCNorton** [comment]
@lorenzoriano all feedback addressed!

File: fiftyone/utils/sam2.py (hunk 1 above)
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py:262]: Added warning in 689a5fe58 when field missing.

File: fiftyone/utils/sam2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py]: 👍 done in 45e45c8ba - added bounds clamping.

File: fiftyone/utils/sam2.py (hunk 3 above)
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py]: addressed in 45e45c8ba.

File: fiftyone/utils/sam2.py (hunk 4 above)
Comment (CharlesCNorton) [reply fiftyone/utils/sam2.py]: also addressed in 45e45c8ba.

File: fiftyone/utils/sam2.py [hunk 5]
```diff
@@ -423,6 +534,27 @@ def _forward_pass_boxes(self, video_reader, sample):
                 mask = np.squeeze(
                     (out_mask_logits[i] > 0.0).cpu().numpy(), axis=0
                 )
+
```
Comment (lorenzoriano) [inline fiftyone/utils/sam2.py:567]: There is significant overlap with the predict_all. Could you do something like:  
"""  
def \_subtract_negative_box_regions(mask, neg_detections, width, height):
    """
    Subtract negative bounding box regions from a mask.
    
    Args:
        mask: numpy array (2D, 3D, or 4D) representing segmentation mask(s)
        neg_detections: fol.Detections containing negative prompt boxes
        width: image/frame width
        height: image/frame height
        
    Returns:
        Modified mask with negative regions zeroed out
    """
    if not neg_detections or not isinstance(neg_detections, fol.Detections):
        return mask
        
    if len(neg_detections.detections) == 0:
        return mask
    
    for neg_det in neg_detections.detections:
        neg_box = neg_det.bounding_box
        box_xyxy = fosam._to_abs_boxes(
            np.array([neg_box]), width, height, chunk_size=1
        )
        box_abs = np.round(box_xyxy.squeeze()).astype(int)
        
        # Clamp to valid bounds
        nx1 = max(0, box_abs[0])
        ny1 = max(0, box_abs[1])
        nx2 = min(width, box_abs[2])
        ny2 = min(height, box_abs[3])
        
        # Only zero out if valid region
        if nx2 > nx1 and ny2 > ny1:
            if mask.ndim == 2:
                # Video model: (H, W)
                mask[ny1:ny2, nx1:nx2] = 0
            elif mask.ndim == 3:
                # Image model: (N, H, W)
                mask[:, ny1:ny2, nx1:nx2] = 0
            elif mask.ndim == 4:
                # Image model: (N, 1, H, W)
                mask[:, 0, ny1:ny2, nx1:nx2] = 0
    
    return mask  
"""  
And use it in both image and video models?

File: tests/unittests/utils/test_sam2.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
+"""
+Tests for fiftyone/utils/sam2.py negative prompt functionality.
```
Comment (lorenzoriano) [inline tests/unittests/utils/test_sam2.py:2]: I think there's more to test here, at least:

- Real mask modifications

- Integration with SAM2 predictor

- Edge cases and error handling  
  
You can ask claude or any other agent to create comprehensive testing for the new functionality

**Comment by CharlesCNorton** [comment]
done in 7cd67a1 👍

**Review by lorenzoriano — APPROVED** [review]
